fix: update oracle spec (#32)
interlay/interbtc-spec: interbtc-spec/docs/source/spec/oracle.rst

Exchange Rate Oracle
====================

> .. note:: This exchange oracle module model that relies on trusted oracle sources. Decentralized oracles are a difficult and open research problem that is outside of the scope of this specification. However, the general interface to get the exchange rate can remain the same even with different constructions.

Reviewer: ```suggestion
.. note:: This exchange rate oracle model relies on trusted oracle sources. Decentralized oracles are a difficult and open research problem that is outside of the scope of this specification. However, the general interface to get the exchange rate can remain the same even with different constructions.
```

> The Exchange Rate Oracle receives a continuous data feed on the from oracles, with information in exchange rates or bitcoin inclusion estimates.  Multiple oracles can be authorized, in which case the 'median' of all unexpired values is used as the actual value. It is not technically the median - when an even number of oracles have submitted values, it does not average the middle two values. Instead, it arbitrarily picks one of them. This is done because this can be done in O(n) rather than in O(n log n). 

Reviewer: ```suggestion
The Exchange Rate Oracle receives a continuous data feed from off-chain oracles, with information in exchange rates or bitcoin inclusion estimates. Multiple oracles can be authorized, in which case the 'median' of all unexpired values is used as the actual value. It is not technically the median - when an even number of oracles have submitted values, it does not average the middle two values. Instead, it arbitrarily picks one of them. This is done because this can be done in O(n) rather than in O(n log n). 
```

The implementation of the oracle client **is not part of this specification**. InterBTC assumes the oracle operates correctly and that the received data is reliable. 

Data Model
~~~~~~~~~~

Enums
-----

OracleKey
.........

Key to indicate a specific value.

.. tabularcolumns:: |l|L|

=======================================  ========================================================================
Discriminant                             Description
=======================================  ========================================================================
``ExchangeRate(CurrencyId)``             Exchange rate against Bitcoin, in e.g. planck per satoshi.
``FeeEstimation(BitcoinInclusionTime)``  Estimate of the bitcoin inclusion fee.
=======================================  ========================================================================

BitcoinInclusionTime
....................

> Indicates the time period for bitcoin inclusion fee estimates.

Reviewer: Perhaps the diff isn't rendering correctly, but this comment is wrong for the table below.

Author: Seems correct for me

*Parameters*
RawValues
.........

* ``oracleId``: the oracle account calling this function.
* ``rate``: the fixed point exchange rate.
Maps OracleKeys and account ids to raw timestamped values. 

*Events*
RawValuesUpdated
................

* :ref:`setExchangeRateEvent`
Maps OracleKey to a boolean value that indicates that a new value has been received that has not yet been included in the aggregate.

*Preconditions*
AuthorizedOracles
.................

* The function call MUST be signed by ``oracleId``.
* The BTC Parachain status in the :ref:`security` component MUST NOT be ``SHUTDOWN:2``.
* The oracle MUST be authorized.
Maps oracle ``accountId`` to the oracle's name. The presence of an account id in this map indicates that the account is authorized to feed values.

*Postconditions*

* The ``ExchangeRate`` MUST be set to the provided ``rate``.
* The ``LastExchangeRateTime`` MUST be updated to the current time.
* If the status in :ref:`security` is ``ERROR:1``, the system MUST be set to ``RUNNING:0``.
Functions
~~~~~~~~~

.. _setBtcTxFeesPerByte:
.. _feedValues:

> setBtcTxFeesPerByte

Reviewer: Not sure what benefit we can from changing the events actually since it doesn't affect the function calling - but does require a re-write in stats.

Author: It was an attempt to keep the API as close as possible to orml-oracle. While it is unlikely that oracle providers depend on the events, it is not impossible, so I chose to change it

*Postconditions*

For each ``(oracleKey, value)`` pair,

* ``RawValuesUpdated[oracleKey]`` MUST be set to true
* ``RawValues[oracleKey]`` MUST be set to a ``TimeStamped`` values, where,

  * ``TimeStamped.timestamp`` MUST be the current time,
  * ``TimeStamped.value`` MUST be ``value``.

.. _getPrice:

getPrice
---------------

Returns the latest medianized value for the given key, as calculated from the received external data sources.

Specification
.............

*Function Signature*

``getPrice(oracleKey)``

*Parameters*

* ``oracleKey``: the key for which the value should be returned

*Preconditions*

> * ``EchangeRate[oracleKey]`` MUST NOT be ``None``. That is, sufficient oracles must have submitted unexpired values.

Reviewer: ```suggestion
* ``ExchangeRate[oracleKey]`` MUST NOT be ``None``. That is, sufficient oracles must have submitted unexpired values.
```

*Postconditions*

* MUST return the fixed point value for the given key.

.. _oracle_onInitialize:

onInitialize
---------------

This function is called at the start of every block. When new values have been submitted, or when old values expire, this function update the aggregate value.

Specification
.............

*Function Signature*

``onInitialize()``

*Postconditions*

* If ``RawValuesUpdated`` is empty, i.e., ``feedValues`` was not yet called since the initialization of the parachain, then the ``OracleOffline`` MUST be set in the :ref:`security` pallet.
* For each ``(oracleKey, updated)`` in ``RawValuesUpdated``, if ``updated`` is true, or the current time is greater than ``ValidUntil[oracle]``,

  * ``RawValuesUpdated[oracleKey]`` MUST be set to false
  * ``ExchangeRate[oracleKey]`` MUST be set to the middle value of the sorted list of unexpired values from ``RawValues[oracleKey]``. If there are an even number, one MAY be arbitrarily picked.
  * ``ValidUntil[oracleKey]`` MUST be set to ``MaxDelay`` plus the minimum timestamp from the unexpired values in ``RawValues[oracleKey]``.

> .. TODO: recover_from_oracle_offline

Reviewer: I think we can reference the security module here right?

Author: Oracle offline detection is still a work in progress
